src/engine/Settings.cs

```diff
@@ -395,7 +455,106 @@ private void CopySettings(Settings settings)
             if (!property.CanWrite)
                 continue;
 
-            property.SetValue(this, property.GetValue(settings));
+            // Since the properties we want to copy are SettingValue generics we use the IAssignableSetting
+            // interface and AssignFrom method to convert the property to the correct concrete class.
+            var setting = property.GetValue(this) as IAssignableSetting;
+
+            setting?.AssignFrom(property.GetValue(settings));
+        }
+    }
+
+    /// <summary>
+    ///   Wrapper class for settings options containing the value and a delegate that provides a callback for
+    ///   when the value is changed.
+    /// </summary>
+    public class SettingValue<TValueType> : IAssignableSetting
+    {
+        private TValueType value;
+
+        public SettingValue(TValueType value)
+        {
+            this.value = value;
+        }
+
+        public event SettingValueChangedDelegate<TValueType> OnChanged;
+
+        public TValueType Value
+        {
+            get => value;
+            set
+            {
+                if (!Equals(this.value, value))
+                {
+                    this.value = value;
+
+                    OnChanged?.Invoke(value);
+                }
+            }
+        }
+
+        public static implicit operator TValueType(SettingValue<TValueType> value)
```

Comment:
Would it also be possible to have an assign operator that takes in TValueType to simplify the assignment of values as well?

Reply:
If I did I'd have to handle the assignment in the property because just implicitly converting from say, a bool, ends up replacing it without keeping the delegate. So I'd have to manually copy the value for each property set.

Reply:
That's true. Is it possible to have explicit constructors in C# to not make that happen?
If that is possible it should make the assignment operator not be confused with creating a new instance.

Reply:
I can't think of a method to do that since the constructor only works for new objects, so it'd still end up assigning it. I mean you could do something like settingValue = new SettingValue<bool>(settingValue, newValue); or feed the delegate in as a parameter, but that's even more convoluted than just selecting the value.

Reply:
After googling it, it seems this is just not possible in C# so `.Value = ` needs to stay.